server/src/main/java/org/elasticsearch/cluster/metadata/MetaDataIndexStateService.java

```diff
@@ -349,29 +371,40 @@ private void sendVerifyShardBeforeCloseRequest(final IndexShardRoutingTable shar
     /**
      * Step 3 - Move index states from OPEN to CLOSE in cluster state for indices that are ready for closing.
      */
-    static ClusterState closeRoutingTable(final ClusterState currentState, final Map<Index, AcknowledgedResponse> results) {
+    static ClusterState closeRoutingTable(final ClusterState currentState,
+                                          final Map<Index, ClusterBlock> blockedIndices,
+                                          final Map<Index, AcknowledgedResponse> results) {
         final MetaData.Builder metadata = MetaData.builder(currentState.metaData());
         final ClusterBlocks.Builder blocks = ClusterBlocks.builder().blocks(currentState.blocks());
         final RoutingTable.Builder routingTable = RoutingTable.builder(currentState.routingTable());
 
         final Set<String> closedIndices = new HashSet<>();
         for (Map.Entry<Index, AcknowledgedResponse> result : results.entrySet()) {
             final Index index = result.getKey();
+            final boolean acknowledged = result.getValue().isAcknowledged();
             try {
+                if (acknowledged == false) {
+                    logger.debug("closing index {} failed", index);
+                    continue;
+                }
                 final IndexMetaData indexMetaData = metadata.getSafe(index);
-                if (indexMetaData.getState() != IndexMetaData.State.CLOSE) {
-                    if (result.getValue().isAcknowledged()) {
-                        logger.debug("closing index {} succeed, removing index routing table", index);
-                        metadata.put(IndexMetaData.builder(indexMetaData).state(IndexMetaData.State.CLOSE));
-                        routingTable.remove(index.getName());
-                        closedIndices.add(index.getName());
-                    } else {
-                        logger.debug("closing index {} failed, removing index block because: {}", index, result.getValue());
-                        blocks.removeIndexBlock(index.getName(), INDEX_CLOSED_BLOCK);
-                    }
-                } else {
-                    logger.debug("index {} has been closed since it was blocked before closing, ignoring", index);
+                if (indexMetaData.getState() == IndexMetaData.State.CLOSE) {
+                    logger.debug("closing index {} succeed  but index is already closed", index);
+                    assert currentState.blocks().hasIndexBlock(index.getName(), INDEX_CLOSED_BLOCK);
+                    continue;
                 }
+                final ClusterBlock closingBlock = blockedIndices.get(index);
+                if (currentState.blocks().hasIndexBlock(index.getName(), closingBlock) == false) {
+                    logger.debug("closing index {} succeed but block has been removed in the mean time", index);
+                    continue;
```

Comment:
why has it succeeded here? The index is not closed and the block has disappeared, so it must have been reopened in the meanwhile. This should be counted as a failure, not a success.

Reply:
Well, the message in the log is confusing. What succeeded here is the verification of shards using the block that has now disappeared. I changed to change the log message.

> This should be counted as a failure, not a success.

I agree. I pushed a change that combines the results of the TransportVerifyShardBeforeCloseAction and the cluster state changes in order to determine if the current close action has indeed closed the index.